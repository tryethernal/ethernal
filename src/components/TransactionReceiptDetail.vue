<template>
  <div class="receipt-tab-content">
    <v-card variant="outlined">
      <v-card-item>
        <v-card-title class="text-subtitle-1 font-weight-bold">Transaction Receipt</v-card-title>
      </v-card-item>
      <v-card-text class="pa-0">
        <v-list density="compact">
          <template v-for="(value, key) in formattedReceipt" :key="key">
            <v-list-item>
              <template v-slot:prepend>
                <div class="text-subtitle-2 font-weight-medium text-grey-darken-1" style="width: 220px;">
                  {{ formatKey(key) }}:
                </div>
              </template>
              <v-list-item-title class="text-body-2">
                {{ formatValue(value) }}
              </v-list-item-title>
            </v-list-item>
          </template>
        </v-list>
      </v-card-text>
    </v-card>
  </div>
</template>

<script setup>
import { computed } from 'vue';
import * as ethers from 'ethers';

const props = defineProps({
  receipt: {
    type: Object,
    required: true
  }
});

// Format receipt for display
const formattedReceipt = computed(() => {
  if (!props.receipt) return {};
  
  // Filter out some properties we don't want to display
  const { logs, ...filteredReceipt } = props.receipt;
  
  return filteredReceipt;
});

// Helper function to format keys
const formatKey = (key) => {
  // Convert camelCase to Title Case with spaces
  return key
    .replace(/([A-Z])/g, ' $1')
    .replace(/^./, str => str.toUpperCase());
};

// Helper function to format values
const formatValue = (value) => {
  if (value === null || value === undefined) return '-';
  
  // Check if it's a BigNumber or can be parsed as one
  try {
    if (typeof value === 'string' && value.startsWith('0x')) {
      if (value.length > 42) return value; // Full hex string
      const num = ethers.BigNumber.from(value);
      return num.toString();
    }
  } catch (e) {
    // Not a BigNumber, continue with other formatting
  }
  
  // Format boolean values
  if (typeof value === 'boolean') {
    return value ? 'Yes' : 'No';
  }
  
  // Format numbers with commas
  if (typeof value === 'number' || (typeof value === 'string' && !isNaN(value))) {
    return Number(value).toLocaleString();
  }
  
  return value.toString();
};
</script>

<style scoped>
.receipt-tab-content :deep(.v-list-item) {
  min-height: 48px;
  padding-top: 8px;
  padding-bottom: 8px;
}
</style> 