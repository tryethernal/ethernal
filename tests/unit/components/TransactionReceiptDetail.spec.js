import TransactionReceiptDetail from '@/components/TransactionReceiptDetail.vue';

describe('TransactionReceiptDetail.vue', () => {
    // Test that the component properly displays and formats all receipt fields
    it('Should display transaction receipt details', async () => {
        const mockReceipt = {
            blockHash: '0x123456789',
            blockNumber: 12345,
            transactionHash: '0x987654321',
            status: true,
            gasUsed: '21000',
            effectiveGasPrice: '0x1234',
            cumulativeGasUsed: 50000,
            from: '0xabcdef123456',
            to: '0x654321fedcba'
        };

        const wrapper = mount(TransactionReceiptDetail, {
            props: {
                receipt: mockReceipt
            }
        });

        expect(wrapper.html()).toMatchSnapshot();
    });

    // Test that the component handles null/undefined values gracefully
    it('Should handle empty or null values', async () => {
        const mockReceipt = {
            blockHash: null,
            blockNumber: undefined,
            transactionHash: '0x987654321',
            status: false
        };

        const wrapper = mount(TransactionReceiptDetail, {
            props: {
                receipt: mockReceipt
            }
        });

        expect(wrapper.html()).toMatchSnapshot();
    });

    // Test that long hex strings are preserved and not converted to numbers
    it('Should format large hex numbers', async () => {
        const mockReceipt = {
            value: '0x1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef'
        };

        const wrapper = mount(TransactionReceiptDetail, {
            props: {
                receipt: mockReceipt
            }
        });

        expect(wrapper.html()).toMatchSnapshot();
    });
}); 