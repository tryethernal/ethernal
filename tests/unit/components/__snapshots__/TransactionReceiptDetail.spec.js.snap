// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TransactionReceiptDetail.vue > Should display transaction receipt details 1`] = `
"<div data-v-5a38f98c="" class="receipt-tab-content">
  <div data-v-5a38f98c="" class="v-card v-theme--light v-card--density-default v-card--variant-outlined">
    <!---->
    <div class="v-card__loader">
      <div class="v-progress-linear v-theme--light v-locale--is-ltr" style="top: 0px; height: 0px; --v-progress-linear-height: 2px;" role="progressbar" aria-hidden="true" aria-valuemin="0" aria-valuemax="100">
        <!---->
        <div class="v-progress-linear__background"></div>
        <div class="v-progress-linear__buffer" style="width: 0%;"></div>
        <transition-stub name="fade-transition" appear="false" persisted="false" css="true">
          <div class="v-progress-linear__indeterminate">
            <div class="v-progress-linear__indeterminate long"></div>
            <div class="v-progress-linear__indeterminate short"></div>
          </div>
        </transition-stub>
        <!---->
      </div>
    </div>
    <!---->
    <!---->
    <div data-v-5a38f98c="" class="v-card-item">
      <!---->
      <div class="v-card-item__content">
        <!---->
        <!---->
        <div data-v-5a38f98c="" class="v-card-title text-subtitle-1 font-weight-bold">Transaction Receipt</div>
      </div>
      <!---->
    </div>
    <div data-v-5a38f98c="" class="v-card-text pa-0">
      <div data-v-5a38f98c="" class="v-list v-theme--light v-list--density-compact v-list--one-line" tabindex="0" role="listbox">
        <div data-v-5a38f98c="" class="v-list-item v-theme--light v-list-item--density-compact v-list-item--one-line rounded-0 v-list-item--variant-text">
          <!----><span class="v-list-item__underlay"></span>
          <div class="v-list-item__prepend">
            <div data-v-5a38f98c="" class="text-subtitle-2 font-weight-medium text-grey-darken-1" style="width: 220px;">Block Hash: </div>
            <div class="v-list-item__spacer"></div>
          </div>
          <div class="v-list-item__content" data-no-activator="">
            <!---->
            <!---->
            <div data-v-5a38f98c="" class="v-list-item-title text-body-2">4886718345</div>
          </div>
          <!---->
        </div>
        <div data-v-5a38f98c="" class="v-list-item v-theme--light v-list-item--density-compact v-list-item--one-line rounded-0 v-list-item--variant-text">
          <!----><span class="v-list-item__underlay"></span>
          <div class="v-list-item__prepend">
            <div data-v-5a38f98c="" class="text-subtitle-2 font-weight-medium text-grey-darken-1" style="width: 220px;">Block Number: </div>
            <div class="v-list-item__spacer"></div>
          </div>
          <div class="v-list-item__content" data-no-activator="">
            <!---->
            <!---->
            <div data-v-5a38f98c="" class="v-list-item-title text-body-2">12,345</div>
          </div>
          <!---->
        </div>
        <div data-v-5a38f98c="" class="v-list-item v-theme--light v-list-item--density-compact v-list-item--one-line rounded-0 v-list-item--variant-text">
          <!----><span class="v-list-item__underlay"></span>
          <div class="v-list-item__prepend">
            <div data-v-5a38f98c="" class="text-subtitle-2 font-weight-medium text-grey-darken-1" style="width: 220px;">Transaction Hash: </div>
            <div class="v-list-item__spacer"></div>
          </div>
          <div class="v-list-item__content" data-no-activator="">
            <!---->
            <!---->
            <div data-v-5a38f98c="" class="v-list-item-title text-body-2">40926266145</div>
          </div>
          <!---->
        </div>
        <div data-v-5a38f98c="" class="v-list-item v-theme--light v-list-item--density-compact v-list-item--one-line rounded-0 v-list-item--variant-text">
          <!----><span class="v-list-item__underlay"></span>
          <div class="v-list-item__prepend">
            <div data-v-5a38f98c="" class="text-subtitle-2 font-weight-medium text-grey-darken-1" style="width: 220px;">Status: </div>
            <div class="v-list-item__spacer"></div>
          </div>
          <div class="v-list-item__content" data-no-activator="">
            <!---->
            <!---->
            <div data-v-5a38f98c="" class="v-list-item-title text-body-2">Yes</div>
          </div>
          <!---->
        </div>
        <div data-v-5a38f98c="" class="v-list-item v-theme--light v-list-item--density-compact v-list-item--one-line rounded-0 v-list-item--variant-text">
          <!----><span class="v-list-item__underlay"></span>
          <div class="v-list-item__prepend">
            <div data-v-5a38f98c="" class="text-subtitle-2 font-weight-medium text-grey-darken-1" style="width: 220px;">Gas Used: </div>
            <div class="v-list-item__spacer"></div>
          </div>
          <div class="v-list-item__content" data-no-activator="">
            <!---->
            <!---->
            <div data-v-5a38f98c="" class="v-list-item-title text-body-2">21,000</div>
          </div>
          <!---->
        </div>
        <div data-v-5a38f98c="" class="v-list-item v-theme--light v-list-item--density-compact v-list-item--one-line rounded-0 v-list-item--variant-text">
          <!----><span class="v-list-item__underlay"></span>
          <div class="v-list-item__prepend">
            <div data-v-5a38f98c="" class="text-subtitle-2 font-weight-medium text-grey-darken-1" style="width: 220px;">Effective Gas Price: </div>
            <div class="v-list-item__spacer"></div>
          </div>
          <div class="v-list-item__content" data-no-activator="">
            <!---->
            <!---->
            <div data-v-5a38f98c="" class="v-list-item-title text-body-2">4660</div>
          </div>
          <!---->
        </div>
        <div data-v-5a38f98c="" class="v-list-item v-theme--light v-list-item--density-compact v-list-item--one-line rounded-0 v-list-item--variant-text">
          <!----><span class="v-list-item__underlay"></span>
          <div class="v-list-item__prepend">
            <div data-v-5a38f98c="" class="text-subtitle-2 font-weight-medium text-grey-darken-1" style="width: 220px;">Cumulative Gas Used: </div>
            <div class="v-list-item__spacer"></div>
          </div>
          <div class="v-list-item__content" data-no-activator="">
            <!---->
            <!---->
            <div data-v-5a38f98c="" class="v-list-item-title text-body-2">50,000</div>
          </div>
          <!---->
        </div>
        <div data-v-5a38f98c="" class="v-list-item v-theme--light v-list-item--density-compact v-list-item--one-line rounded-0 v-list-item--variant-text">
          <!----><span class="v-list-item__underlay"></span>
          <div class="v-list-item__prepend">
            <div data-v-5a38f98c="" class="text-subtitle-2 font-weight-medium text-grey-darken-1" style="width: 220px;">From: </div>
            <div class="v-list-item__spacer"></div>
          </div>
          <div class="v-list-item__content" data-no-activator="">
            <!---->
            <!---->
            <div data-v-5a38f98c="" class="v-list-item-title text-body-2">188900967593046</div>
          </div>
          <!---->
        </div>
        <div data-v-5a38f98c="" class="v-list-item v-theme--light v-list-item--density-compact v-list-item--one-line rounded-0 v-list-item--variant-text">
          <!----><span class="v-list-item__underlay"></span>
          <div class="v-list-item__prepend">
            <div data-v-5a38f98c="" class="text-subtitle-2 font-weight-medium text-grey-darken-1" style="width: 220px;">To: </div>
            <div class="v-list-item__spacer"></div>
          </div>
          <div class="v-list-item__content" data-no-activator="">
            <!---->
            <!---->
            <div data-v-5a38f98c="" class="v-list-item-title text-body-2">111339007564986</div>
          </div>
          <!---->
        </div>
      </div>
    </div>
    <!---->
    <!----><span class="v-card__underlay"></span>
  </div>
</div>"
`;

exports[`TransactionReceiptDetail.vue > Should format large hex numbers 1`] = `
"<div data-v-5a38f98c="" class="receipt-tab-content">
  <div data-v-5a38f98c="" class="v-card v-theme--light v-card--density-default v-card--variant-outlined">
    <!---->
    <div class="v-card__loader">
      <div class="v-progress-linear v-theme--light v-locale--is-ltr" style="top: 0px; height: 0px; --v-progress-linear-height: 2px;" role="progressbar" aria-hidden="true" aria-valuemin="0" aria-valuemax="100">
        <!---->
        <div class="v-progress-linear__background"></div>
        <div class="v-progress-linear__buffer" style="width: 0%;"></div>
        <transition-stub name="fade-transition" appear="false" persisted="false" css="true">
          <div class="v-progress-linear__indeterminate">
            <div class="v-progress-linear__indeterminate long"></div>
            <div class="v-progress-linear__indeterminate short"></div>
          </div>
        </transition-stub>
        <!---->
      </div>
    </div>
    <!---->
    <!---->
    <div data-v-5a38f98c="" class="v-card-item">
      <!---->
      <div class="v-card-item__content">
        <!---->
        <!---->
        <div data-v-5a38f98c="" class="v-card-title text-subtitle-1 font-weight-bold">Transaction Receipt</div>
      </div>
      <!---->
    </div>
    <div data-v-5a38f98c="" class="v-card-text pa-0">
      <div data-v-5a38f98c="" class="v-list v-theme--light v-list--density-compact v-list--one-line" tabindex="0" role="listbox">
        <div data-v-5a38f98c="" class="v-list-item v-theme--light v-list-item--density-compact v-list-item--one-line rounded-0 v-list-item--variant-text">
          <!----><span class="v-list-item__underlay"></span>
          <div class="v-list-item__prepend">
            <div data-v-5a38f98c="" class="text-subtitle-2 font-weight-medium text-grey-darken-1" style="width: 220px;">Value: </div>
            <div class="v-list-item__spacer"></div>
          </div>
          <div class="v-list-item__content" data-no-activator="">
            <!---->
            <!---->
            <div data-v-5a38f98c="" class="v-list-item-title text-body-2">0x1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef</div>
          </div>
          <!---->
        </div>
      </div>
    </div>
    <!---->
    <!----><span class="v-card__underlay"></span>
  </div>
</div>"
`;

exports[`TransactionReceiptDetail.vue > Should handle empty or null values 1`] = `
"<div data-v-5a38f98c="" class="receipt-tab-content">
  <div data-v-5a38f98c="" class="v-card v-theme--light v-card--density-default v-card--variant-outlined">
    <!---->
    <div class="v-card__loader">
      <div class="v-progress-linear v-theme--light v-locale--is-ltr" style="top: 0px; height: 0px; --v-progress-linear-height: 2px;" role="progressbar" aria-hidden="true" aria-valuemin="0" aria-valuemax="100">
        <!---->
        <div class="v-progress-linear__background"></div>
        <div class="v-progress-linear__buffer" style="width: 0%;"></div>
        <transition-stub name="fade-transition" appear="false" persisted="false" css="true">
          <div class="v-progress-linear__indeterminate">
            <div class="v-progress-linear__indeterminate long"></div>
            <div class="v-progress-linear__indeterminate short"></div>
          </div>
        </transition-stub>
        <!---->
      </div>
    </div>
    <!---->
    <!---->
    <div data-v-5a38f98c="" class="v-card-item">
      <!---->
      <div class="v-card-item__content">
        <!---->
        <!---->
        <div data-v-5a38f98c="" class="v-card-title text-subtitle-1 font-weight-bold">Transaction Receipt</div>
      </div>
      <!---->
    </div>
    <div data-v-5a38f98c="" class="v-card-text pa-0">
      <div data-v-5a38f98c="" class="v-list v-theme--light v-list--density-compact v-list--one-line" tabindex="0" role="listbox">
        <div data-v-5a38f98c="" class="v-list-item v-theme--light v-list-item--density-compact v-list-item--one-line rounded-0 v-list-item--variant-text">
          <!----><span class="v-list-item__underlay"></span>
          <div class="v-list-item__prepend">
            <div data-v-5a38f98c="" class="text-subtitle-2 font-weight-medium text-grey-darken-1" style="width: 220px;">Block Hash: </div>
            <div class="v-list-item__spacer"></div>
          </div>
          <div class="v-list-item__content" data-no-activator="">
            <!---->
            <!---->
            <div data-v-5a38f98c="" class="v-list-item-title text-body-2">-</div>
          </div>
          <!---->
        </div>
        <div data-v-5a38f98c="" class="v-list-item v-theme--light v-list-item--density-compact v-list-item--one-line rounded-0 v-list-item--variant-text">
          <!----><span class="v-list-item__underlay"></span>
          <div class="v-list-item__prepend">
            <div data-v-5a38f98c="" class="text-subtitle-2 font-weight-medium text-grey-darken-1" style="width: 220px;">Block Number: </div>
            <div class="v-list-item__spacer"></div>
          </div>
          <div class="v-list-item__content" data-no-activator="">
            <!---->
            <!---->
            <div data-v-5a38f98c="" class="v-list-item-title text-body-2">-</div>
          </div>
          <!---->
        </div>
        <div data-v-5a38f98c="" class="v-list-item v-theme--light v-list-item--density-compact v-list-item--one-line rounded-0 v-list-item--variant-text">
          <!----><span class="v-list-item__underlay"></span>
          <div class="v-list-item__prepend">
            <div data-v-5a38f98c="" class="text-subtitle-2 font-weight-medium text-grey-darken-1" style="width: 220px;">Transaction Hash: </div>
            <div class="v-list-item__spacer"></div>
          </div>
          <div class="v-list-item__content" data-no-activator="">
            <!---->
            <!---->
            <div data-v-5a38f98c="" class="v-list-item-title text-body-2">40926266145</div>
          </div>
          <!---->
        </div>
        <div data-v-5a38f98c="" class="v-list-item v-theme--light v-list-item--density-compact v-list-item--one-line rounded-0 v-list-item--variant-text">
          <!----><span class="v-list-item__underlay"></span>
          <div class="v-list-item__prepend">
            <div data-v-5a38f98c="" class="text-subtitle-2 font-weight-medium text-grey-darken-1" style="width: 220px;">Status: </div>
            <div class="v-list-item__spacer"></div>
          </div>
          <div class="v-list-item__content" data-no-activator="">
            <!---->
            <!---->
            <div data-v-5a38f98c="" class="v-list-item-title text-body-2">No</div>
          </div>
          <!---->
        </div>
      </div>
    </div>
    <!---->
    <!----><span class="v-card__underlay"></span>
  </div>
</div>"
`;
